### ISSUE TITLE
testWidgets causes incorrect file/line/column for tests (returns location of the test() call inside flutter_test)

### ISSUE THREAD

**Issue by DanTup** [flutter/flutter#18195]
The test framework returns the location for a test so that an editor can jump to it when the user clicks on it. For tests that use `testWidgets` it returns the location of the `test` call inside `testWidgets` so it won't take the user to where they expect.

Example:

```js
{
	"test": {
		"id": 3,
		"name": "Changing locale",
		"suiteID": 0,
		"groupIDs": [
			2
		],
		"line": 60,
		"column": 18,
		"url": "package:flutter_test/src/widget_tester.dart"
	},
	"type": "testStart",
	"time": 21952
}
```

I need to use this information to navigate to the test when a user clicks in the tree on the left. Currently I ignore any `package:` URIs so it just does nothing, but it makes the test runner much less useful.

![screen shot 2018-06-05 at 10 42 33 am](https://user-images.githubusercontent.com/1078012/40968350-44ea54fe-68ad-11e8-906e-ea1af2ca42f3.png)

**Comment by DanTup** [comment]
@jakemac53  @natebosch Is this something that'd be better in the test repo? I'm not sure where it can be fixed (but I do think we need to do something).

**Comment by jakemac53** [comment]
I wonder if we could look at the stack and report the last line that came from the original test file?

**Comment by DanTup** [comment]
Sounds reasonable to me (though probably it's the first frame that matches that you want, if you're starting at the top). It'd still fall down if the user had their own wrapper over `test` that happened to be in the same file, but I think it'd catch the huge majority of cases.

I just tested in Android Studio and seems that suffers because of this too - you can only jump to a file from the suite node, you can't navigate from a test. Surprised it hasn't come up before!

**Comment by jakemac53** [comment]
Can you test out with this branch https://github.com/dart-lang/test/tree/report-original-test-location? 

This adds additional fields, `root_line`, `root_column`, and `root_url` to the json reporter.

If this covers your use case it shouldn't be to much work to add some tests and land it, although I haven't thought through the implications of adding fields to this protocol to much.

@natebosch wdyt?

**Comment by natebosch** [comment]
Seems sensible to me.

https://github.com/dart-lang/test/blob/report-original-test-location/doc/json_reporter.md is the docs for the format, the assumption is that it is safe to add new fields.

**Comment by DanTup** [comment]
I'm struggling to get Flutter to use this version... I've set `path:` in my pubspec for both my checked out `flutter_test` and in my `stocks` example app, and also made it use the local path for `flutter_test`. The `.packages` files all updated and show paths to the checked out version; yet I'm not seeing this (I even changed the `protocolVersion` in the local copy to check, and it's not getting the updated version).

I'll test without Flutter (just putting a helper in another file), but it won't be quite as accurate a test.

**Comment by DanTup** [comment]
In a non-Flutter test, this looks perfect :-)

```js
{
"line": 4,
"column": 3,
"url": "file:///Users/dantup/Dev/Dart-Code/test/test_projects/hello_world/test/helpers.dart",
"root_line": 6,
"root_column": 3,
"root_package_url": "file:///Users/dantup/Dev/Dart-Code/test/test_projects/hello_world/test/main.dart"
}
```

**Comment by jakemac53** [comment]
> In a non-Flutter test, this looks perfect :-)

Flutter does have their own test platform, and they might have forked the json reporter as well - so that probably has to be updated in a similar fashion, but it should be trivial.

**Comment by DanTup** [comment]
`flutter_test` has a dependency on `test` and a comment about pinning to a specific version, so I assumed it used it; but there is the possibility it's doing some stuff (like the reporter) itself. That might also explain why we can debug tests using the VM using the json reporter in Flutter but not the VM. In any case, once it's done here we can patch the Flutter one to be the same, I can look into that.

**Comment by DanTup** [comment]
From discussions with Devon, I don't think it has its own reporter. So if this lands here I'll try revving the version in Flutter locally and re-test (and if no good, I'll track down what needs fixing in Flutter).

**Comment by jakemac53** [comment]
landed in 0.12.40

**Comment by DanTup** [comment]
Flutter has been updated and this all works great now 👍
